plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt

```diff
@@ -114,9 +118,189 @@ class K2ElementActionsFactory : JvmElementActionsFactory() {
+        } ?: return emptyList()
+
+        return listOf(ChangeAnnotationAction(annotationEntry, attributeIndex, request, text, familyName))
+    }
+
+    //TODO: this is copy paste from KotlinElementActionsFactory, maybe we should extract it to common place?
+    // or implement this for K1 and K2?
+
+    private class ChangeAnnotationAction(
+        annotationEntry: KtAnnotationEntry,
+        private val attributeIndex: Int,
+        private val request: AnnotationAttributeRequest,
+        @IntentionName private val text: String,
+        @IntentionFamilyName private val familyName: String
+    ) : IntentionAction {
+
+        private val pointer: SmartPsiElementPointer<KtAnnotationEntry> = annotationEntry.createSmartPointer()
+        private val qualifiedName: String = annotationEntry.toLightAnnotation()?.qualifiedName ?: throw IllegalStateException("r")
+
+        override fun startInWriteAction(): Boolean = true
+
+        override fun getFamilyName(): String = familyName
+
+        override fun getText(): String = text
+
+        override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean = pointer.element != null
+
+        override fun generatePreview(project: Project, editor: Editor, file: PsiFile): IntentionPreviewInfo {
+            invokeImpl(PsiTreeUtil.findSameElementInCopy(pointer.element, file), project)
+            return IntentionPreviewInfo.DIFF
+        }
+
+        override fun invoke(project: Project, editor: Editor?, file: PsiFile) {
+            val annotationEntry = pointer.element ?: return
+            invokeImpl(annotationEntry, project)
+        }
+
+        private fun invokeImpl(annotationEntry: KtAnnotationEntry, project: Project) {
+            val facade = JavaPsiFacade.getInstance(annotationEntry.project)
+            val language = facade.findClass(qualifiedName, annotationEntry.resolveScope)?.language
+            val dummyAnnotationRequest = annotationRequest(qualifiedName, request)
+            val psiFactory = KtPsiFactory(project)
+            val dummyAnnotationText = '@' + renderAnnotation(dummyAnnotationRequest, psiFactory) { language == KotlinLanguage.INSTANCE }
+            val dummyArgumentList = psiFactory.createAnnotationEntry(dummyAnnotationText).valueArgumentList!!
+            val argumentList = annotationEntry.valueArgumentList
+
+            if (argumentList == null) {
+                annotationEntry.add(dummyArgumentList)
+                //ShortenReferences.DEFAULT.process(annotationEntry)
+                return
+            }
+
+            when (language) {
+                JavaLanguage.INSTANCE -> changeJava(annotationEntry, argumentList, dummyArgumentList)
+                KotlinLanguage.INSTANCE -> changeKotlin(annotationEntry, argumentList, dummyArgumentList)
+                else -> changeKotlin(annotationEntry, argumentList, dummyArgumentList)
+            }
+            //ShortenReferences.DEFAULT.process(annotationEntry)
```

Comment:
there is `shortenReference` function or `org.jetbrains.kotlin.idea.base.codeInsight.ShortenReferencesFacility` for the common code